### DISCUSSION TITLE
Dynamically load components if/when needed

### DISCUSSION THREAD

**Discussion by ianhempsell** [vercel/next.js#46553] [Help] [answered]
Hi all... I'm new to JS dev (finally breaking away from PHP/WordPress) and I'm stuck with a concept.
Hopefully, someone can point me in the right direction.

I have an array of components from an API:

$components = ['Text', 'Image', 'Video', 'Slider'];

For each item within the array, There is a basic component that returns it's name within the root directory /components/

On my Page component, I would like to import/require only the components that are in the $components array - this needs to be dynamic as the array could have 1 || 30 items within it - as I'm making a 'page-builder' for a marketing website.

So... something like this within the Page component:

return (
components.map((item, index){
        const Component = require(`../../components/${item}`).default;

        return <Component key={`${item}-${index}`}>
    });
);

So the rendered result would be:

<Text />
<Image />
<Video />
<Slider />

Any help is really appreciated.

BTW, I'm using Next 13 if that helps.

**Comment by ciruz** [comment, marked as answer]
Hey,

one solution with ``next/dynamic``, could look like this:

```javascript
import dynamic from "next/dynamic";

const TextComponent = dynamic(() => import("./components/text"));
const ImageComponent = dynamic(() => import("./components/image"));


const Components = {
  Text: TextComponent ,
  Image: ImageComponent ,
  //...
};

const dataFromApi = ['Text', 'Image', 'Video', 'Slider'];

 return (
    <>
      {dataFromApi.map((component, index) => {
          return React.createElement(Components[component], {
            key: index
          });
        })}
    </>
  );



```

**Reply by ianhempsell** [reply]
Hey Ciruz,

I've tried that already but that relies on 'manually' adding/including/importing all components before the default export and I was trying to 'automate' the inclusions based on the data in the API endpoint.

Thanks anyway.

**Reply by ciruz** [reply]
Okay, I would prefer this solution since you have more control what get's loaded and maybe have different component names too, so you have no conflicts.

But, if you want to be 100% dynamic, you can use something like this:

```javascript
import dynamic from "next/dynamic";

const dataFromApi = ['Text', 'Image', 'Video', 'Slider'];

 return (
    <>
      {dataFromApi.map((component, index) => {
        const DynamicComponent = dynamic(() => import(`../components/${component}`));
        return <DynamicComponent key={index} />;
      })}
    </>
  );
```

I just tested it with a clean next 13 installation, it works.

**Comment by ianhempsell** [comment]
Oh my days... Thanks Ciruz that has worked... BUT...
I had actually tried this method days ago and I just get this error:

`Error: Cannot access LoadableContext._context on the server. You cannot dot into a client module from a server component. You can only pass the imported name through.`

I was using the turbo flag and as soon as I removed it, the solutions worked.

You wouldn't happen to know why this wouldn't work with --turbo would you?
So I can regognise the mistake in the future.

Thanks BTW for you help. Ian.

**Reply by ianhempsell** [reply]
Actually, reading the Turbo docs, next/dynamic is not yet supported but plans are to fully support when stable.

**Reply by ciruz** [reply]
Okay, a good information, thank you
